File: src/windows/Maple.WinUI/MapleUI/Organisms/MuiInfoPanel.cs
using System;
using System.Collections.Generic;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using Maple.UI.Atoms;

namespace Maple.UI
{
    /// <summary>
    /// Maple.UI Info Panel organism (unified-component-catalog.md §4.3,
    /// "Info Panel" row: "Full asset metadata", built from Label-Value
    /// Grid, Histogram, Keyword Row, Rating &amp; Flags, Inline Rename
    /// Field) — the read/light-edit metadata tab of the Inspector Panel:
    /// filename (rename-in-place), rating/flag, the live RGB histogram,
    /// EXIF as a Label-Value Grid, and keywords.
    /// </summary>
    public sealed class MuiInfoPanel : ContentControl
    {
        public static readonly DependencyProperty FilenameProperty =
            DependencyProperty.Register(nameof(Filename), typeof(string), typeof(MuiInfoPanel),
                new PropertyMetadata(string.Empty, (d, _) => ((MuiInfoPanel)d).Rebuild()));

        public static readonly DependencyProperty RatingProperty =
            DependencyProperty.Register(nameof(Rating), typeof(int), typeof(MuiInfoPanel),
                new PropertyMetadata(0, (d, e) => ((MuiInfoPanel)d)._ratingFlags.Rating = (int)e.NewValue));

        public static readonly DependencyProperty FlagProperty =
            DependencyProperty.Register(nameof(Flag), typeof(MuiRatingFlagState), typeof(MuiInfoPanel),
                new PropertyMetadata(MuiRatingFlagState.None, (d, e) => ((MuiInfoPanel)d)._ratingFlags.Flag = (MuiRatingFlagState)e.NewValue));

        public static readonly DependencyProperty RedValuesProperty =
            DependencyProperty.Register(nameof(RedValues), typeof(IReadOnlyList<double>), typeof(MuiInfoPanel),
                new PropertyMetadata(null, (d, e) => ((MuiInfoPanel)d)._histogram.RedValues = (IReadOnlyList<double>?)e.NewValue));

        public static readonly DependencyProperty GreenValuesProperty =
            DependencyProperty.Register(nameof(GreenValues), typeof(IReadOnlyList<double>), typeof(MuiInfoPanel),
                new PropertyMetadata(null, (d, e) => ((MuiInfoPanel)d)._histogram.GreenValues = (IReadOnlyList<double>?)e.NewValue));

        public static readonly DependencyProperty BlueValuesProperty =
            DependencyProperty.Register(nameof(BlueValues), typeof(IReadOnlyList<double>), typeof(MuiInfoPanel),
                new PropertyMetadata(null, (d, e) => ((MuiInfoPanel)d)._histogram.BlueValues = (IReadOnlyList<double>?)e.NewValue));

        public static readonly DependencyProperty ExifRowsProperty =
            DependencyProperty.Register(nameof(ExifRows), typeof(IReadOnlyList<MuiLabelValueRow>), typeof(MuiInfoPanel),
                new PropertyMetadata(null, (d, e) => ((MuiInfoPanel)d)._exifGrid.Rows = (IReadOnlyList<MuiLabelValueRow>?)e.NewValue));

        public static readonly DependencyProperty KeywordsProperty =
            DependencyProperty.Register(nameof(Keywords), typeof(IReadOnlyList<MuiChip>), typeof(MuiInfoPanel),
                new PropertyMetadata(null, (d, e) => ((MuiInfoPanel)d)._keywordRow.Keywords = (IReadOnlyList<MuiChip>?)e.NewValue));

        public string Filename { get => (string)GetValue(FilenameProperty); set => SetValue(FilenameProperty, value); }
        public int Rating { get => (int)GetValue(RatingProperty); set => SetValue(RatingProperty, value); }
        public MuiRatingFlagState Flag { get => (MuiRatingFlagState)GetValue(FlagProperty); set => SetValue(FlagProperty, value); }
        public IReadOnlyList<double>? RedValues { get => (IReadOnlyList<double>?)GetValue(RedValuesProperty); set => SetValue(RedValuesProperty, value); }
        public IReadOnlyList<double>? GreenValues { get => (IReadOnlyList<double>?)GetValue(GreenValuesProperty); set => SetValue(GreenValuesProperty, value); }
        public IReadOnlyList<double>? BlueValues { get => (IReadOnlyList<double>?)GetValue(BlueValuesProperty); set => SetValue(BlueValuesProperty, value); }
        public IReadOnlyList<MuiLabelValueRow>? ExifRows { get => (IReadOnlyList<MuiLabelValueRow>?)GetValue(ExifRowsProperty); set => SetValue(ExifRowsProperty, value); }
        public IReadOnlyList<MuiChip>? Keywords { get => (IReadOnlyList<MuiChip>?)GetValue(KeywordsProperty); set => SetValue(KeywordsProperty, value); }

        public event EventHandler<string>? Renamed;
        public event EventHandler<int>? RatingChanged;
        public event EventHandler<MuiRatingFlagState>? FlagChanged;
        public event EventHandler<string>? KeywordRemoved;
        public event EventHandler<string>? KeywordAdded;

        private readonly StackPanel _root = new() { Orientation = Orientation.Vertical, Spacing = 20 };
        private readonly MuiInlineRenameField _renameField = new();
        private readonly MuiRatingFlags _ratingFlags = new();
        private readonly MuiHistogram _histogram = new() { PlotWidth = 260, PlotHeight = 90 };
        private readonly MuiLabelValueGrid _exifGrid = new();
        private readonly MuiKeywordRow _keywordRow = new() { AddPlaceholder = "Add keyword" };

        public MuiInfoPanel()
        {
            var header = new StackPanel { Orientation = Orientation.Horizontal, Spacing = 12 };
            header.Children.Add(_renameField);
            header.Children.Add(_ratingFlags);

            _root.Children.Add(header);
            _root.Children.Add(SectionLabel("Histogram"));
            _root.Children.Add(_histogram);
            _root.Children.Add(SectionLabel("Metadata"));
            _root.Children.Add(_exifGrid);
            _root.Children.Add(SectionLabel("Keywords"));
            _root.Children.Add(_keywordRow);
            Content = _root;

            _renameField.Renamed += (_, name) => { Filename = name; Renamed?.Invoke(this, name); };
            _ratingFlags.RatingChanged += (_, value) => { Rating = value; RatingChanged?.Invoke(this, value); };
            _ratingFlags.FlagChanged += (_, value) => { Flag = value; FlagChanged?.Invoke(this, value); };
            _keywordRow.Removed += (_, id) => KeywordRemoved?.Invoke(this, id);
            _keywordRow.Added += (_, label) => KeywordAdded?.Invoke(this, label);

            Rebuild();
        }

        private static MuiText SectionLabel(string text) => new() { Text = text, Variant = MuiTextVariant.Eyebrow, ColorRole = MuiTextColorRole.Muted };

        private void Rebuild()
        {
            _renameField.Value = Filename;
            _renameField.AccessibleLabel = $"Rename {Filename}";
        }
    }
}

File: src/apple/Packages/MapleUI/Sources/MapleUI/Organisms/MuiInfoPanel.swift
// MuiInfoPanel.swift — Maple UI Organisms · Inspectors & panels
// (unified-component-catalog.md §4.3). Full asset metadata, built from
// Label-Value Grid, Histogram, Keyword Row, Rating & Flags, Inline Rename
// Field. `loading` gates the whole body behind a centered Spinner.
// `metadata` always renders — Label-Value Grid already handles an empty
// array gracefully — so there's no separate metadata-empty state to build.

import SwiftUI

public struct MuiInfoPanelHistogram: Sendable {
    public let r: [Int]
    public let g: [Int]
    public let b: [Int]

    public init(r: [Int], g: [Int], b: [Int]) {
        self.r = r
        self.g = g
        self.b = b
    }
}

public struct MuiInfoPanel: View {
    public let loading: Bool
    @Binding public var filename: String
    public let metadata: [MuiLabelValueRow]
    public let histogram: MuiInfoPanelHistogram?
    public let keywords: [MuiChip]
    @Binding public var rating: Int
    @Binding public var flag: MuiRatingFlagState
    public let renamed: ((String) -> Void)?
    public let keywordAdded: ((String) -> Void)?
    public let keywordRemoved: ((String) -> Void)?

    public init(
        loading: Bool = false,
        filename: Binding<String>,
        metadata: [MuiLabelValueRow],
        histogram: MuiInfoPanelHistogram? = nil,
        keywords: [MuiChip],
        rating: Binding<Int> = .constant(0),
        flag: Binding<MuiRatingFlagState> = .constant(.none),
        renamed: ((String) -> Void)? = nil,
        keywordAdded: ((String) -> Void)? = nil,
        keywordRemoved: ((String) -> Void)? = nil
    ) {
        self.loading = loading
        self._filename = filename
        self.metadata = metadata
        self.histogram = histogram
        self.keywords = keywords
        self._rating = rating
        self._flag = flag
        self.renamed = renamed
        self.keywordAdded = keywordAdded
        self.keywordRemoved = keywordRemoved
    }

    public var body: some View {
        Group {
            if loading {
                MuiSpinner(placement: .centered, label: "Loading metadata")
            } else {
                ScrollView {
                    VStack(alignment: .leading, spacing: MuiTokens.spacingMd) {
                        MuiInlineRenameField(value: $filename, accessibilityLabel: "Filename", renamed: renamed)

                        if let histogram {
                            MuiHistogram(r: histogram.r, g: histogram.g, b: histogram.b)
                        }

                        MuiRatingFlags(rating: $rating, flag: $flag)

                        MuiKeywordRow(keywords: keywords, removed: keywordRemoved, added: keywordAdded)

                        MuiLabelValueGrid(rows: metadata)
                    }
                    .padding(MuiTokens.spacingMd)
                }
            }
        }
    }
}

#Preview("MuiInfoPanel — Populated") {
    struct Demo: View {
        @State private var filename = "IMG_0042.dng"
        @State private var rating = 4
        @State private var flag: MuiRatingFlagState = .pick
        var body: some View {
            MuiInfoPanel(
                filename: $filename,
                metadata: [
                    MuiLabelValueRow(label: "Camera", value: "Sony A7 IV"),
                    MuiLabelValueRow(label: "Lens", value: "24-70mm f/2.8"),
                    MuiLabelValueRow(label: "ISO", value: "400"),
                ],
                histogram: MuiInfoPanelHistogram(r: (0..<32).map { $0 * 3 }, g: (0..<32).map { $0 * 2 }, b: (0..<32).map { 64 - $0 }),
                keywords: [MuiChip(id: "1", label: "Sunset"), MuiChip(id: "2", label: "Landscape")],
                rating: $rating,
                flag: $flag
            )
            .frame(width: 280, height: 480)
            .background(MuiTokens.bg)
        }
    }
    return Demo()
}

#Preview("MuiInfoPanel — Loading") {
    MuiInfoPanel(filename: .constant(""), metadata: [], keywords: [])
        .frame(width: 280, height: 200)
        .background(MuiTokens.bg)
}
